cmd/connqc/server.go

```diff
@@ -22,34 +25,53 @@ func runServer(c *cli.Context) error {
 	readTimeout := c.Duration(flagReadTimeout)
 	writeTimeout := c.Duration(flagWriteTimeout)
 
-	sigSrv := connqc.NewServer(bufferSize, readTimeout, writeTimeout, log)
+	srv := connqc.NewServer(bufferSize, readTimeout, writeTimeout, log)
 
-	srv, err := tcp.NewServer(sigSrv)
+	tcpSrv, err := tcp.NewServer(srv)
+	if err != nil {
+		return err
+	}
+
+	udpSrv, err := udp.NewServer(srv)
 	if err != nil {
 		return err
 	}
 
 	addr := c.String(flagAddr)
+
+	grp := sync.WaitGroup{}
+	grp.Add(2)
+
 	log.Info("Starting server",
 		lctx.Str("addr", addr),
 		lctx.Int("buffer_size", bufferSize),
 		lctx.Duration("read_timeout", readTimeout),
 		lctx.Duration("write_timeout", writeTimeout),
 	)
 	go func() {
-		if err := srv.Listen(addr); err != nil && !errors.Is(err, tcp.ErrServerClosed) {
-			log.Error("Server error", lctx.Error("error", err))
+		defer grp.Done()
+
+		if err := tcpSrv.Listen(ctx, addr); err != nil && !errors.Is(err, net.ErrClosed) {
+			log.Error("Server error", lctx.Str("protocol", "tcp"), lctx.Err(err))
+			return
+		}
+		log.Info("TCP server stopped")
+	}()
+	go func() {
+		defer grp.Done()
+
+		if err := udpSrv.Listen(ctx, addr); err != nil && !errors.Is(err, net.ErrClosed) {
+			log.Error("Server error", lctx.Str("protocol", "udp"), lctx.Err(err))
+			return
 		}
+		log.Info("UDP server stopped")
 	}()
```

Comment:
When it's like that I usually prefer to do the `Add` in the relevant place, so that if the code gets removed or moved around, the `Add` does too:

```go
	grp := sync.WaitGroup{}

	log.Info("Starting server",
		lctx.Str("addr", addr),
		lctx.Int("buffer_size", bufferSize),
		lctx.Duration("read_timeout", readTimeout),
		lctx.Duration("write_timeout", writeTimeout),
	)
	go func() {
		grp.Add(1)
		defer grp.Done()

		if err := tcpSrv.Listen(ctx, addr); err != nil && !errors.Is(err, net.ErrClosed) {
			log.Error("Server error", lctx.Str("protocol", "tcp"), lctx.Err(err))
			return
		}
		log.Info("TCP server stopped")
	}()
	go func() {
		grp.Add(1)
		defer grp.Done()

		if err := udpSrv.Listen(ctx, addr); err != nil && !errors.Is(err, net.ErrClosed) {
			log.Error("Server error", lctx.Str("protocol", "udp"), lctx.Err(err))
			return
		}
		log.Info("UDP server stopped")
	}()
```

WDYT?

Reply:
Linter complained about a potential race, which make sense, so I put it right before.

Reply:
Right my bad. Do you mean you put it right before the call to `go func`? This is how https://gobyexample.com/waitgroups seems to recommend it

Reply:
See https://github.com/nitrado/connqc/pull/20/commits/3ffbb48ed82cc0be4b7786fdbe3ee0ffbc1bf4f3. Basically

```go
grp.Add(1)
go func() {
  defer grp.Done()
  // ..
}
```